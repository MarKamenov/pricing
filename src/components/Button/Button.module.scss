.button {
    padding: .5rem 1rem;
    border-radius: .3rem;
    font-weight: 600;
    font-size: 0.875rem;
    cursor: pointer;
    transition: background-color 0.2s ease, transform 0.1s ease;
    border: none;
    outline: none;

    &:active {
        transform: scale(0.98);
    }

    &.fullWidth {
        width: 100%;
    }

    &.disabled {
        cursor: not-allowed;

        &:active {
            transform: none;
        }
    }
}

.primary {
    background-color: var(--button-primary-bck);
    color: var(--button-primary-text-clr);

    &:hover:not(:disabled) {
        background-color: var(--button-primary-hover-bck);
        box-shadow: 0px 4px 14px var(--button-shadow-clr);
    }

    &:active:not(:disabled) {
        background-color: var(--button-primary-hover-bck);
    }

    &.disabled {
        background-color: var(--button-primary-disabled-bck);
    }
}

.secondary {
    background-color: var(--button-secondary-bck);
    border: 1.5px solid var(--violet-700);
    color: var(--button-secondary-text-clr);

    &:hover:not(:disabled) {
        background-color: var(--button-secondary-hover-bck);
        box-shadow: 0px 4px 14px var(--button-shadow-clr);
    }

    &:active:not(:disabled) {
        background-color: var(--button-secondary-hover-bck);
    }

    &.disabled {
        background-color: var(--button-secondary-disabled-bck);
    }
}

.small {
    font-size: var(--font-size-sm);
    padding: var(--spacing-xs) var(--spacing-sm);
}

.medium {
    font-size: var(--font-size-md);
    padding: 0.75rem var(--spacing-md);
}

.large {
    font-size: var(--font-size-lg);
    padding: var(--spacing-sm) var(--spacing-lg);
}