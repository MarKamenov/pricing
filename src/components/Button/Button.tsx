import { ReactNode } from 'react';
import { cn } from '../../lib';
import buttonStyles from './Button.module.scss';

export interface ButtonProps {
    children: ReactNode
    variant?: "primary" | "secondary"
    onClick: () => void;
    size?: "small" | "medium" | "large";
    className?: string;
    ariaLabel: string;
    fullWidth?: boolean
    disabled?: boolean
}

export const Button = ({ ariaLabel, size = "small", children, className, variant = 'primary', onClick, fullWidth = false, disabled = false }: ButtonProps) => {
    const buttonClass = cn(buttonStyles.button, buttonStyles[variant], buttonStyles[size], fullWidth ? buttonStyles.fullWidth : "", disabled ? buttonStyles.disabled : "", className)

    const handleKeyDown = (event: React.KeyboardEvent<HTMLButtonElement>) => {
        if ((event.key === "Enter" || event.key === " ") && !disabled) {
            event.preventDefault();
            onClick();
        }
    };

    return (
        <button type='button' aria-label={ariaLabel} className={buttonClass} onClick={onClick} onKeyDown={handleKeyDown} disabled={disabled}>
            {children}
        </button>
    )
}