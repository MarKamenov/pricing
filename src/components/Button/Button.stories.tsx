import type { Meta, StoryObj } from "@storybook/react";
import { Button } from "./Button";

const meta: Meta<typeof Button> = {
    title: "Components/Button",
    component: Button,
    argTypes: {
        onClick: { action: "clicked" },
        variant: { control: "radio", options: ["primary", "secondary"] },
        size: { control: "radio", options: ["small", "medium", "large"] },
        fullWidth: { control: "boolean" },
        disabled: { control: "boolean" },
    },
    tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Primary: Story = {
    args: {
        children: "Primary Button",
        variant: "primary",
        ariaLabel: "Primary button",
        onClick: () => console.log("Primary button clicked"),
    },
};

export const Secondary: Story = {
    args: {
        children: "Secondary Button",
        variant: "secondary",
        ariaLabel: "Secondary button",
        onClick: () => console.log("Secondary button clicked"),
    },
};

export const Disabled: Story = {
    args: {
        children: "Disabled Button",
        disabled: true,
        ariaLabel: "Disabled button",
        onClick: () => console.log("Disabled button clicked"),
    },
};

export const FullWidth: Story = {
    args: {
        children: "Full Width Button",
        fullWidth: true,
        ariaLabel: "Full width button",
        onClick: () => console.log("Full width button clicked"),
    },
};
