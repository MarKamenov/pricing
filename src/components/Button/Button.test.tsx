import { fireEvent, render, screen } from "@testing-library/react"
import { Button } from './Button';

describe('Button component', () => {
    const mockOnClick = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should render component', () => {
        const { container } = render(<Button ariaLabel='label' onClick={mockOnClick}>Button text</Button>)

        expect(container).toBeDefined();
    });

    test('should match snapshot', () => {
        const { container } = render(<Button ariaLabel='label' onClick={mockOnClick}>Button text</Button>)

        expect(container).toMatchSnapshot();
    });

    test("renders correctly with given props", () => {
        render(
            <Button ariaLabel="Test Button" onClick={mockOnClick}>
                Click Me
            </Button>
        );

        const button = screen.getByRole("button", { name: /test button/i });
        expect(button).toBeInTheDocument();
        expect(button).toHaveTextContent("Click Me");
        expect(button).not.toBeDisabled();
    });

    test("calls onClick when clicked", () => {
        render(
            <Button ariaLabel="Clickable Button" onClick={mockOnClick}>
                Click Me
            </Button>
        );

        const button = screen.getByRole("button", { name: /clickable button/i });
        fireEvent.click(button);
        expect(mockOnClick).toHaveBeenCalledTimes(1);
    });

    test("does not call onClick when disabled", () => {
        render(
            <Button ariaLabel="Disabled Button" onClick={mockOnClick} disabled>
                Cannot Click
            </Button>
        );

        const button = screen.getByRole("button", { name: /disabled button/i });
        expect(button).toBeDisabled();
        fireEvent.click(button);
        expect(mockOnClick).not.toHaveBeenCalled();
    });

    test("is accessible via keyboard (Enter key)", () => {
        render(
            <Button ariaLabel="Keyboard Accessible Button" onClick={mockOnClick}>
                Press Enter
            </Button>
        );

        const button = screen.getByRole("button", { name: /keyboard accessible button/i });
        button.focus();
        expect(button).toHaveFocus();

        fireEvent.keyDown(button, { key: "Enter", code: "Enter" });

        expect(mockOnClick).toHaveBeenCalledTimes(1);
    });
});

