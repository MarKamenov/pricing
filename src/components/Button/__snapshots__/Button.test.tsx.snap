// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should match snapshot 1`] = `
<div>
  <button
    aria-label="label"
    class="button primary small"
    type="button"
  >
    Button text
  </button>
</div>
`;
